platform/logging/LoggerFactory.ts

```diff
@@ -0,0 +1,109 @@
+    let context = namespace.join('.');
+    if (this.loggers.has(context)) {
+      return this.loggers.get(context)!;
+    }
+
+    let loggerConfig;
+    if (this.config) {
+      loggerConfig = this.getLoggerConfigByContext(this.config, context);
+    } else {
+      // If we don't have config yet, use `buffered` appender that will store all logged messages in the memory
+      // until the config is ready.
+      loggerConfig = { level: LogLevel.All, appenders: [BUFFERED_APPENDER_KEY] };
+    }
+
+    this.loggers.set(context,
+      new LoggerAdapter(
+        new BaseLogger(
+          namespace,
+          loggerConfig.level,
+          loggerConfig.appenders.map((appenderKey) => this.appenders.get(appenderKey)!)
+        )
+      )
+    );
+
+    return this.loggers.get(context)!;
+  }
+
+  updateConfig(config: LoggingConfig) {
+    // TODO: Should we support `config.isEqual()` to avoid updating the config if it's still the same?
+    this.config = config;
+
+    // Before we dispose appenders, let's check whether we have any `buffered` appender that may need to be
+    // flushed via newly configured appenders.
+    const bufferedLogRecords = this.appenders.has(BUFFERED_APPENDER_KEY) ?
+      [...(<BufferAppender>this.appenders.get(BUFFERED_APPENDER_KEY)!).buffer] :
+      [];
+
+    this.updateAppendersFromConfig(config);
+    this.updateLoggersFromConfig(config);
+
+    // Re-log all buffered log records with newly configured appenders.
+    for (const logRecord of bufferedLogRecords) {
+      this.get(logRecord.context.join('.')).log(logRecord);
+    }
+  }
+
+  close(): void {
+    for (const appender of this.appenders.values()) {
+      appender.close();
+    }
+
+    this.appenders.clear();
+    this.loggers.clear();
+  }
+
+  private updateAppendersFromConfig(config: LoggingConfig) {
+    for (const appender of this.appenders.values()) {
+      appender.close();
+    }
+
+    this.appenders.clear();
+
+    // Initialize new log appenders.
+    for (const [appenderKey, appenderConfig] of config.appenders.entries()) {
+      this.appenders.set(appenderKey, appenderConfig.createAppender());
+    }
+  }
+
+  private updateLoggersFromConfig(config: LoggingConfig) {
+    for (const [loggerKey, loggerAdapter] of this.loggers.entries()) {
+      const loggerConfig = this.getLoggerConfigByContext(config, loggerKey);
+      loggerAdapter.logger = new BaseLogger(
+        // TODO: Do we need to store namespace as an array at all?
+        loggerKey.split('.'),
+        loggerConfig.level,
+        loggerConfig.appenders.map((appenderKey) => this.appenders.get(appenderKey)!)
+      );
+    }
+  }
+
+  private getLoggerConfigByContext(config: LoggingConfig, context: string) {
+    // If we don't have dedicated configuration for specified context, let's use config from `root` context,
+    // that is always should be defined (enforced by configuration schema).
+    return config.loggers.get(config.loggers.has(context) ? context : 'root')!;
```

Comment:
Ugh, if we use `.` as context separator it seems we can't really define a schema for it (since we expand these properties into nested objects and in the example above `server` schema should allow `level`, `appenders` and `foo` properties), unless I'm missing something. If that's true, maybe we can use another separator (eg. `server:foo` or `server::foo`)?